File: src/Blokemon.CardGen/Domain/CardProfile.cs
namespace Blokemon.CardGen.Domain;

/// <summary>The printed physical description of a collectible.</summary>
/// <param name="Feet">The whole feet of the length.</param>
/// <param name="Inches">The remaining inches of the length.</param>
/// <param name="Pounds">The weight in pounds.</param>
public sealed record Stature(int Feet, int Inches, int Pounds)
{
    /// <summary>The length as printed on the identity strip.</summary>
    /// <returns>The printed length.</returns>
    public string PrintedLength() => Inches == 0 ? $"{Feet}'" : $"{Feet}' {Inches}\"";

    /// <summary>The weight as printed on the identity strip.</summary>
    /// <returns>The printed weight.</returns>
    public string PrintedWeight() => $"{Pounds} lbs.";
}

/// <summary>The species epithet and stature printed on the identity strip.</summary>
/// <param name="Subtype">The epithet naming the collectible's species.</param>
/// <param name="Stature">The printed length and weight.</param>
public sealed record CardProfile(string Subtype, Stature Stature)
{
    /// <summary>The identity strip line.</summary>
    /// <returns>The printed identity.</returns>
    public string PrintedIdentity() =>
        $"{Subtype} Blokemon. Length: {Stature.PrintedLength()}, Weight: {Stature.PrintedWeight()}";
}

File: src/Blokemon.CardGen/Domain/CardProfile.fs
namespace Blokemon.CardGen.Domain

/// The printed physical description of a collectible.
type Stature =
    {
        /// The whole feet of the length.
        Feet: int

        /// The remaining inches of the length.
        Inches: int

        /// The weight in pounds.
        Pounds: int
    }

    /// The length as printed on the identity strip.
    member this.PrintedLength() =
        if this.Inches = 0 then
            $"{this.Feet}'"
        else
            $"{this.Feet}' {this.Inches}\""

    /// The weight as printed on the identity strip.
    member this.PrintedWeight() = $"{this.Pounds} lbs."

/// The species epithet and stature printed on the identity strip.
type CardProfile =
    {
        /// The epithet naming the collectible's species.
        Subtype: string

        /// The printed length and weight.
        Stature: Stature
    }

    /// The identity strip line.
    member this.PrintedIdentity() =
        let length = this.Stature.PrintedLength()
        let weight = this.Stature.PrintedWeight()
        $"{this.Subtype} Blokemon. Length: {length}, Weight: {weight}"
